dusty/systems/docker/__init__.py

```diff
@@ -26,6 +28,39 @@ def get_dusty_images():
 def get_dusty_container_name(service_name):
     return 'dusty_{}_1'.format(service_name)
 
+def registry_from_image(image_name):
+    """Returns the Docker registry host associated with
+    a given image name."""
+    if '/' not in image_name: # official image
+        return 'index.docker.io'
+    prefix = image_name.split('/')[0]
+    if '.' not in prefix: # user image on official repository, e.g. thieman/clojure
+        return 'index.docker.io'
+    return prefix # custom registry, e.g. gamechanger.io/lox
+
+@memoized
+def get_authed_registries():
+    """Reads the local Docker client config for the current user
+    and returns all registries to which the user may be logged in.
+    This is intended to be run client-side, not by the daemon."""
+    result = set()
+    if not os.path.exists(constants.DOCKER_CONFIG_PATH):
+        return result
+    config = json.load(open(constants.DOCKER_CONFIG_PATH, 'r'))
+    for registry in config.get('auths', {}).iterkeys():
+        try:
+            parsed = urlparse(registry)
+        except Exception:
+            log_to_client('Error parsing registry {} from Docker config, will skip this registry').format(registry)
+        # This logic assumes the auth is either of the form
+        # gamechanger.io (no scheme, no path after host) or
+        # of the form https://index.docker.io/v1/ (scheme,
+        # netloc parses correctly, additional path does not matter).
+        # These are the formats I saw in my personal config file,
+        # not sure what other formats it might accept.
+        result.add(parsed.netloc) if parsed.netloc else result.add(parsed.path)
+    return result
+
```

Comment:
Rationale was I thought it should live in the Docker submodule somewhere since it's dealing with a) the Docker client's config file and b) Docker's various formats for image naming. Could split it out into a separate submodule within `/docker/` but definitely think it belongs in there somewhere.